[dtensor][1/n] refactor op dispatch logic to reduce overhead

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107307
* #107306
* __->__ #107305

This PR is the first change of a series of refactors to the op dispatch logic to:
1. remove the redundant logic in the op dispatch, simplify the error
checking
2. reduce the number of tree_map/tree_flatten/unflatten needed to reduce
the overhead coming from those operations
3. remove the CachedShardingPropagator by using lru_cache from functools
directly, this makes it not only helps TP, but general DTensor
operations could be faster!
4. change the view ops behavior by inplace changing the op_schema, which
is dangerous for sharding prop caching, model the view op as one type
of resharding too
5. enrich output sharding to include whether the op needs redistribute
so that we don't need explicit op schema comparison to know it.

This should help with further reducing the CPU overhead, benchmark
results:
before (without this change), aten.addmm latency: 0.476ms
![Screenshot 2023-08-16 at 10 46 26 AM](https://github.com/pytorch/pytorch/assets/9443650/7692e6c1-1936-4c7f-bf9c-6c8c9b8f6c76)

after (with this change), aten.addmm latency: 0.341ms
![Screenshot 2023-08-16 at 11 05 49 AM](https://github.com/pytorch/pytorch/assets/9443650/15a53f0b-7a95-444e-ab2f-3ee0ad2fa47f)


overall one layer of mlp time reduced from 13.535 -> 9.665ms

Apart from overhead reduction, this PR simplifies the op dispatching logic and the resharding logic (more refactor needed to make things more clean, which will be done in later PRs)

wanchaol committed:
[dtensor][1/n] refactor op dispatch logic to reduce overhead

This PR is the first change of a series of refactors to the op dispatch logic to:
1. remove the redundant logic in the op dispatch, simplify the error
checking
2. reduce the number of tree_map/tree_flatten/unflatten needed to reduce
the overhead coming from those operations
3. remove the CachedShardingPropagator by using lru_cache from functools
directly
4. change the view ops behavior by inplace changing the op_schema, which
is dangerous for sharding prop caching, model the view op as one type
of resharding too
5. enrich output sharding to include whether the op needs redistribute
so that we don't need explicit op schema comparison to know it.

This should help with further reducing the CPU overhead, benchmark
results coming soon

By refactoring the dispatch logic, it could possibly enable us to have
more features later (i.e. add IrregularShard placement easier)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/n] refactor op dispatch logic to reduce overhead"


This PR is the first change of a series of refactors to the op dispatch logic to:
1. remove the redundant logic in the op dispatch, simplify the error
checking
2. reduce the number of tree_map/tree_flatten/unflatten needed to reduce
the overhead coming from those operations
3. remove the CachedShardingPropagator by using lru_cache from functools
directly, this makes it not only helps TP, but general DTensor
operations could be faster!
4. change the view ops behavior by inplace changing the op_schema, which
is dangerous for sharding prop caching, model the view op as one type
of resharding too
5. enrich output sharding to include whether the op needs redistribute
so that we don't need explicit op schema comparison to know it.

This should help with further reducing the CPU overhead, benchmark
results coming soon

By refactoring the dispatch logic, it could possibly enable us to have
more features later (i.e. add IrregularShard placement easier)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/n] refactor op dispatch logic to reduce overhead"


This PR is the first change of a series of refactors to the op dispatch logic to:
1. remove the redundant logic in the op dispatch, simplify the error
checking
2. reduce the number of tree_map/tree_flatten/unflatten needed to reduce
the overhead coming from those operations
3. remove the CachedShardingPropagator by using lru_cache from functools
directly, this makes it not only helps TP, but general DTensor
operations could be faster!
4. change the view ops behavior by inplace changing the op_schema, which
is dangerous for sharding prop caching, model the view op as one type
of resharding too
5. enrich output sharding to include whether the op needs redistribute
so that we don't need explicit op schema comparison to know it.

This should help with further reducing the CPU overhead, benchmark
results:
before (without this change), aten.addmm latency: 0.476ms
![Screenshot 2023-08-16 at 10 46 26 AM](https://github.com/pytorch/pytorch/assets/9443650/7692e6c1-1936-4c7f-bf9c-6c8c9b8f6c76)

after (with this change), aten.addmm latency: 0.341ms
![Screenshot 2023-08-16 at 11 05 49 AM](https://github.com/pytorch/pytorch/assets/9443650/15a53f0b-7a95-444e-ab2f-3ee0ad2fa47f)


overall one layer of mlp time reduced from 13.535 -> 9.665ms


[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/n] refactor op dispatch logic to reduce overhead"


This PR is the first change of a series of refactors to the op dispatch logic to:
1. remove the redundant logic in the op dispatch, simplify the error
checking
2. reduce the number of tree_map/tree_flatten/unflatten needed to reduce
the overhead coming from those operations
3. remove the CachedShardingPropagator by using lru_cache from functools
directly, this makes it not only helps TP, but general DTensor
operations could be faster!
4. change the view ops behavior by inplace changing the op_schema, which
is dangerous for sharding prop caching, model the view op as one type
of resharding too
5. enrich output sharding to include whether the op needs redistribute
so that we don't need explicit op schema comparison to know it.

This should help with further reducing the CPU overhead, benchmark
results:
before (without this change), aten.addmm latency: 0.476ms
![Screenshot 2023-08-16 at 10 46 26 AM](https://github.com/pytorch/pytorch/assets/9443650/7692e6c1-1936-4c7f-bf9c-6c8c9b8f6c76)

after (with this change), aten.addmm latency: 0.341ms
![Screenshot 2023-08-16 at 11 05 49 AM](https://github.com/pytorch/pytorch/assets/9443650/15a53f0b-7a95-444e-ab2f-3ee0ad2fa47f)


overall one layer of mlp time reduced from 13.535 -> 9.665ms


[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -1,4 +1,5 @@
-from typing import Callable, Dict, Optional, Tuple
+from functools import lru_cache
```
Does this one comes with cache missing rate stats? So that we can debug or see how the cache is working.

wanchaol replied:
added a API to explicitly get the missing rate/hit rate https://github.com/pytorch/pytorch/pull/107305/files#diff-bc0598e52e1f48ecf4f904ebbe5fd206854933871c3ffe305f95885ecbf53edcR4

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -123,45 +116,68 @@ def _operator_dispatch(
-                " torch.Tensor to DTensor before calling distributed operators!"
-            )
 
+    for arg in flat_args_list:
```
Can we also move this to treemap? Not sure how much perf benefit we can get out of it though.

fduwjj replied:
oh got it, good to know!

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -123,45 +116,68 @@ def _operator_dispatch(
+                    )
+            else:
+                mesh = kwarg.device_mesh
+        else:
```
For the completeness of this logic, we also need something like from L139-L143?

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -123,45 +116,68 @@ def _operator_dispatch(
+        op_call,
+        OpSchema(
+            op_call._schema,
+            tree_unflatten(flat_args_schema, args_spec),
+            tree_unflatten(flat_kwargs_schema, kwargs_spec),
```
Is there anyway that we can avoid this unflatten? To me this seems to be unnecessary?

fduwjj replied:
I see. Make sense, if treemap is unavoidable, ideally we only want to do it once.

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -123,45 +116,68 @@ def _operator_dispatch(
+
+    sharding_propagator.propagate(op_info)
+    output_sharding = op_info.output_sharding
+    assert output_sharding is not None
```
nit: maybe add messages here for the assert?

fduwjj commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -216,17 +226,19 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
+            # compute locally with redistribute first if needed
+            assert output_sharding.schema_suggestions is not None
+            suggested_input_schema = output_sharding.schema_suggestions[0]
+            redistribute_local_args(op_info, suggested_input_schema)
```
This name is a little bit confusing to me because the concept `local` to me is more like tensor rather DTensor while redistribute is a unique thing for DTensor. Would it be possible that we can just call it `redistribute_args` or `redistribute_n_update_local_args`? I understand the logic here is to first redistribute dtensor and update local_args so that we pass the correct args to the final Aten ops. But the naming seems to suggest we are doing redistribute directly on top of a Tensor.

fduwjj replied:
After reading your comment below and a second thought... I think this name is fine.

fduwjj commented on torch/distributed/_tensor/redistribute.py:
```diff
@@ -72,19 +73,25 @@ def _decompose_reshard(val: List[_PlacementItem]) -> List[_PlacementItem]:
 
 # Intentionally expose this API to trace ops on local tensors
-def _redistribute_with_local_tensor(
+def redistribute_local_tensor(
```
ditto, like I mentioned in the `dispatch.py`, maybe the old name `redistribute_with_local_tensor` or redistribute_with_local_tensor_updated` is better? because  the name here seems to suggest that we directly redistribute a Tensor, which is somewhat confusing?

wanchaol replied:
hmmm this function/api is indeed directly redistribute a torch.Tensor (i.e. the first arg is the local_tensor) with src/dst dtensor spec, maybe we should add more comment on the API to clarify?

fduwjj replied:
OK, essentially there is no DTensor involved indeed. We just call collectives directly (and with no-autograd). I just think redistributing a tensor is a little confusing. Well since we redistribute a local tensor of a DTensor, this sounds make sense to me. If we can add some comments here, that would be much appreciated.

wanchaol replied:
Yep will add in the follow up PR

fduwjj reviewed (commented):
Nice work!! I have some questions around naming and cache miss stats, also is it possible to change the how much we cache in the lru cache?

fduwjj commented on torch/distributed/_tensor/ops/view_ops.py:
```diff
@@ -618,18 +614,30 @@ def reshape_prop(op_schema: OpSchema) -> OutputSharding:
-                op_schema.args_schema = (
-                    args[:shape_argnum]
+                # compute the local shape from the global shape, then return
+                # a resharding even if we don't really reshard, the only reason
+                # for this type of resharding is to lower the global shape to
+                # local shape
+                local_out_shape = compute_local_shape(
+                    list(global_out_shape), mesh, shard_out
+                )
+
+                suggested_schema = OpSchema(
+                    func_schema=op_schema.func_schema,
+                    args_schema=args[:shape_argnum]
                     + (tuple(local_out_shape),)
-                    + args[shape_argnum + 1 :]
+                    + args[shape_argnum + 1 :],
+                    kwargs_schema=op_schema.kwargs_schema,
+                )
+                return OutputSharding(
+                    output_spec=output_dtensor_spec,
+                    schema_suggestions=[suggested_schema],
+                    needs_redistribute=True,
```
Nice! It would be interesting to see if we still have any cache miss on both MLP mini example and transformer model. For example, detach and index op I observed in the past which also have op_schema changing.

wanchaol committed:
Update on "[dtensor][1/n] refactor op dispatch logic to reduce overhead"


This PR is the first change of a series of refactors to the op dispatch logic to:
1. remove the redundant logic in the op dispatch, simplify the error
checking
2. reduce the number of tree_map/tree_flatten/unflatten needed to reduce
the overhead coming from those operations
3. remove the CachedShardingPropagator by using lru_cache from functools
directly, this makes it not only helps TP, but general DTensor
operations could be faster!
4. change the view ops behavior by inplace changing the op_schema, which
is dangerous for sharding prop caching, model the view op as one type
of resharding too
5. enrich output sharding to include whether the op needs redistribute
so that we don't need explicit op schema comparison to know it.

This should help with further reducing the CPU overhead, benchmark
results:
before (without this change), aten.addmm latency: 0.476ms
![Screenshot 2023-08-16 at 10 46 26 AM](https://github.com/pytorch/pytorch/assets/9443650/7692e6c1-1936-4c7f-bf9c-6c8c9b8f6c76)

after (with this change), aten.addmm latency: 0.341ms
![Screenshot 2023-08-16 at 11 05 49 AM](https://github.com/pytorch/pytorch/assets/9443650/15a53f0b-7a95-444e-ab2f-3ee0ad2fa47f)


overall one layer of mlp time reduced from 13.535 -> 9.665ms

Apart from overhead reduction, this PR simplifies the op dispatching logic and the resharding logic (more refactor needed to make things more clean, which will be done in later PRs)


[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -30,6 +32,7 @@ def __init__(self) -> None:
             OpOverload,
             Callable[[Node, DeviceMesh, Dict[Node, StrategyType]], StrategyType],
         ] = {}
+        self.propagate_op_sharding = lru_cache(None)(self.propagate_op_sharding)  # type: ignore[method-assign]
```
Yep cache hit should happen automatically with the python built-in lru cache

wanchaol committed:
Update on "[dtensor][1/n] refactor op dispatch logic to reduce overhead"


This PR is the first change of a series of refactors to the op dispatch logic to:
1. remove the redundant logic in the op dispatch, simplify the error
checking
2. reduce the number of tree_map/tree_flatten/unflatten needed to reduce
the overhead coming from those operations
3. remove the CachedShardingPropagator by using lru_cache from functools
directly, this makes it not only helps TP, but general DTensor
operations could be faster!
4. change the view ops behavior by inplace changing the op_schema, which
is dangerous for sharding prop caching, model the view op as one type
of resharding too
5. enrich output sharding to include whether the op needs redistribute
so that we don't need explicit op schema comparison to know it.

This should help with further reducing the CPU overhead, benchmark
results:
before (without this change), aten.addmm latency: 0.476ms
![Screenshot 2023-08-16 at 10 46 26 AM](https://github.com/pytorch/pytorch/assets/9443650/7692e6c1-1936-4c7f-bf9c-6c8c9b8f6c76)

after (with this change), aten.addmm latency: 0.341ms
![Screenshot 2023-08-16 at 11 05 49 AM](https://github.com/pytorch/pytorch/assets/9443650/15a53f0b-7a95-444e-ab2f-3ee0ad2fa47f)


overall one layer of mlp time reduced from 13.535 -> 9.665ms

Apart from overhead reduction, this PR simplifies the op dispatching logic and the resharding logic (more refactor needed to make things more clean, which will be done in later PRs)


[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/debug/__init__.py:
```diff
@@ -0,0 +1,10 @@
+from torch.distributed._tensor.api import DTensor
+
+
+def get_sharding_prop_cache_info():
```
Thanks for adding this. It would be nice to add a comment or a reference on Python LRU so that user know what information we can get from this API. (This can be done in a follow-up PR as a BE work)

wanchaol replied:
Yeah sure I'll add comments in a follow up PR :)

fduwjj reviewed (approved):
LGTM and thanks for the refactoring.

wanchaol commented:
@pytorchbot merge -i

huydhn commented:
@pytorchbot drci
